llvm/lib/CodeGen/RegAllocGreedy.cpp

```diff
@@ -321,10 +327,12 @@ unsigned DefaultPriorityAdvisor::getPriority(const LiveInterval &LI) const {
     // Giant live ranges fall back to the global assignment heuristic, which
     // prevents excessive spilling in pathological cases.
     const TargetRegisterClass &RC = *MRI->getRegClass(Reg);
-    bool ForceGlobal = RC.GlobalPriority ||
-                       (!ReverseLocalAssignment &&
-                        (Size / SlotIndex::InstrDist) >
-                            (2 * RegClassInfo.getNumAllocatableRegs(&RC)));
+
+    bool ForceGlobal =
+        RC.GlobalPriority ||
+        (!GreedyForceLocalAssignment && !ReverseLocalAssignment &&
+         (Size / SlotIndex::InstrDist) >
+             (2 * RegClassInfo.getNumAllocatableRegs(&RC)));
```

Comment:
This heuristic is broken and I've really wanted to fix this for a long time. getNumAllocatableRegs does not understand overlapping registers. What this really should be accounting for is number of independently allocatable registers. Can we fix that instead of just adding a flag to skip the heuristic?

Reply:
My plan is to use this flag to skip, and in separate PR introduce TRI.forceGlobalAssignment which gives targets a chance to write their own. Other targets may be fine with the current implementation (backend has limited calls to SlotIndex renumbering, minimal overlapping regs), and universally changing could cause regressions or increase compile time for no reason.

Even if we fix the heuristic, there are still edge cases where we would like to override

Reply:
No, I think this heuristic is just broken and is not a point for customization. It just happens to work for CPUs that don't have dense register files. It should work exactly the same for CPU targets if it were in terms of independently allocatable registers